src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -295,6 +312,23 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
         continue;
       }
 
+      // Check if the scheduling class is at capacity now.
+      if (sched_cls_info.running_tasks.size() >= sched_cls_info.capacity &&
+          work->GetState() == internal::WorkStatus::WAITING) {
+        RAY_LOG(DEBUG) << "Hit cap! time=" << get_time_ms_()
+                       << " next update time=" << sched_cls_info.next_update_time;
+        if (get_time_ms_() < sched_cls_info.next_update_time) {
+          // We're over capacity and it's not time to admit a new task yet.
+          // Calculate the next time we should admit a new task.
+          UpdateSchedulingClassWaitTime(sched_cls_info);
```

Comment:
I'm confused by the logic here. Why do we update the timer each time we see that we haven't reached the timer yet? That seems like it could cause us to hang if we always call ScheduleAndDispatch before the timeout (if this logic is right, can you check this behavior in the unit tests?).

It seems like it should be:
```
if (next time not set) {
  update next time
  break
} else if (get_time > next time) {
  schedule task, reset next time to nil
} else {
  break
}
```

Seems like the first and third cases are getting squashed together into one case now.

Reply:
> That seems like it could cause us to hang if we always call ScheduleAndDispatch before the timeout (if this logic is right, can you check this behavior in the unit tests?).

I'm not sure I understand what you mean, but the update takes the min of the existing wait time and the newly calculated one. Since we're calling ScheduleAndDispatch every 100ms, and the existing test cases take 10+ seconds to timeout, it seems like we must be calling ScheduleAndDispatch before the timeout?

Reply:
I see, I didn't catch the "min" part. I found this confusing because the UpdateSchedulingClassWaitTime makes it sound like the timer will get renewed, but it's actually only getting renewed if it's not already set.

Can you take care of the other comment to inline the code here? I think that will help with the clarity.